syntax = "proto3";

package ollama;

// Represents a single text chunk
message TextChunk {
  string content = 1;      
  int32 order = 2;         
  bool is_last = 3;       
}

// Metadata generated for a chunk
message ChunkMetadata {
  int32 order = 1;             
  string title = 2;            
  string description = 3;      
  repeated string keywords = 4; 
  string topic = 5;            
}

// Request to generate metadata for a single chunk
message GenerateChunkRequest {
  string model = 1;
  TextChunk chunk = 2;
  int32 max_tokens = 3;     
}

// Response with metadata for a single chunk
message GenerateChunkResponse {
  ChunkMetadata metadata = 1;
}

// Request to aggregate multiple chunk metadata into an episode
message AggregateRequest {
  string model = 1;
  repeated ChunkMetadata chunk_metadata = 2; 
  int32 max_tokens = 3;
}

// Response with aggregated episode metadata
message AggregateResponse {
  string title = 1;              
  string description = 2;        
  repeated string keywords = 3;  
  string topic = 4;              
}
  
// Ollama gRPC service
service Ollama {
  rpc GenerateChunk (GenerateChunkRequest) returns (GenerateChunkResponse);
  rpc AggregateChunks (AggregateRequest) returns (AggregateResponse);
}
